Fix for Issue #4

louoso opened the pull request:
Using softValues seems more appropriate. It will be purged when memory is needed, but not on every GC. It also breaks the Strong Value to Weak Key link that kept memory from being reclaimed. Finally, object equals instead of identity will result in fewer cache entries.

Commit:
fixing classInfoCache- A new entry was added for each key instance and values were never purged

advorkovyy commented:
Louis,

I do not see how the strong reference from ClassInfo to ScalaType is broken. SoftReference is much like WeakReference, right? it is just less eager to be GCed.

louoso commented:
Using softValues or weakValues will allow the cache to be purged. The Guava mapMaker javadocs state-

"If soft or weak references were requested, it is possible for a key or value present in the the map to be reclaimed by the garbage collector. If this happens, the entry automatically disappears from the map."

So when the value is reclaimed, the key will also be released. This essentially fixes the issue without having to remove the ScalaType reference from ClassInfo.

The other thing I noticed is this-

"Note: by default, the returned map uses equality comparisons (the equals method) to determine equality for keys or values. However, if weakKeys() or softKeys() was specified, the map uses identity (==) comparisons instead for keys. Likewise, if weakValues() or softValues() was specified, the map uses identity comparisons for values."

This is confusing because it is not using scala's == operator, but in fact java's object equals. So using weakKeys actually results in repeated entries for ScalaType and exacerbates the cache issue. Changing to weakValues or softValues has the same consequence for values, but I don't think there are any calls to the affected methods (contains, replace, etc).

Whether to use weakValues or softValues is really up to you, but when profiling, I felt the cache was much more effective using softValues. The weakValues were expelled on every GC, which in a production setting, can be surprisingly often.

louoso commented:
The strong reference from ClassInfo to ScalaType means that the ScalaType instance is always strongly reachable as long as the ClassInfo is strongly reachable. This means that even though the ScalaType is weakly reachable when used as a key, it will never be gc'd because it is always strongly reachable via the value (ClassInfo). Additionally, the identity equals means that the ScalaType will only work well as a key so long as you are using the same instance, which is not always the case.

advorkovyy commented:
Yes, I see. I would like to merge it into development branch.

Commit:
Fix for this cache never being purged.

The weak key is a Class instance which is always strongly reachable via the class loader.

louoso commented:
Ok, new pull request for development branch.
